src/target/llvm/codegen_llvm.cc

```diff
@@ -215,41 +205,29 @@ void CodeGenLLVM::LinkParameters(const Map<String, LinkedParam> params) {
 
   llvm::BasicBlock* entry = llvm::BasicBlock::Create(*ctx_, "entry", function);
   builder_->SetInsertPoint(entry);
-  std::vector<llvm::Value*> zero_index_list{llvm::ConstantInt::get(t_int32_, 0)};
-  std::vector<llvm::Value*> zero_array_index_list{llvm::ConstantInt::get(t_int32_, 0),
-                                                  llvm::ConstantInt::get(t_int32_, 0)};
-  auto args_array = builder_->CreateBitCast(
-#if TVM_LLVM_VERSION >= 50
-      &function->arg_begin()[0],
+
+  auto getArg = [function](int i) -> llvm::Argument* {
+#if TVM_LLVM_VERSION >= 100
+    return function->getArg(i);
+#elif TVM_LLVM_VERSION >= 50
+    return &function->arg_begin()[i];
 #else
-      &(*(function->arg_begin())),
+    return &*std::next(function->arg_begin(), i);
 #endif
-      llvm::ArrayType::get(t_void_->getPointerTo(GetGlobalAddressSpace()), 1));
-  llvm::Value* sid = builder_->CreateBitCast(
-      builder_->CreateLoad(t_void_->getPointerTo(GetGlobalAddressSpace()),
-                           builder_->CreateInBoundsGEP(args_array, zero_index_list)),
-      t_int64_);
+  };
+
+  llvm::Type* t_int64_p = t_int64_->getPointerTo(GetGlobalAddressSpace());
+  llvm::Value* sid = builder_->CreateLoad(t_int64_, builder_->CreateBitCast(getArg(0), t_int64_p));
+
+  auto ret_tcode = builder_->CreateBitCast(getArg(4), t_int_p);
+  auto ret_value =
+      builder_->CreateBitCast(getArg(3), t_void_p_->getPointerTo(GetGlobalAddressSpace()));
 
   llvm::BasicBlock* default_block = llvm::BasicBlock::Create(*ctx_, "default_block", function);
-  auto ret_types_array = builder_->CreateBitCast(
-#if TVM_LLVM_VERSION >= 50
-      &function->arg_begin()[4],
-#else
-      &(*(std::next(function->arg_begin(), 4))),
-#endif
-      llvm::ArrayType::get(t_int_, 1)->getPointerTo());
-  auto retval_array = builder_->CreateBitCast(
-#if TVM_LLVM_VERSION >= 50
-      &function->arg_begin()[3],
-#else
-      &(*std::next(function->arg_begin(), 3)),
-#endif
-      llvm::ArrayType::get(t_void_->getPointerTo(GetGlobalAddressSpace()), 1)->getPointerTo());
   llvm::SwitchInst* switch_inst = builder_->CreateSwitch(sid, default_block, params.size() + 1);
 
   builder_->SetInsertPoint(default_block);
-  builder_->CreateStore(llvm::ConstantInt::get(t_int_, kTVMNullptr),
-                        builder_->CreateInBoundsGEP(ret_types_array, zero_array_index_list));
+  builder_->CreateStore(llvm::ConstantInt::get(t_int_, kTVMNullptr), ret_tcode);
```

Comment:
i'm really the furthest from an expert at LLVM bitcode, but it seems like this stops treating ret_tcode as `int[]` and merely as the int* it is. is that bad?

Reply:
`int[n]` in LLVM is `[n x i32]`, and it's an aggregate type, i.e. it's the entire array. If you want to load or store an element of it, you need the address of the beginning of it, i.e. `i32*`.  Unlike in C, arrays in LLVM IR are not treated as pointers.

Reply:
right, so given ret_tcode is `int[1]`, should we be representing that in its llvm type?

Reply:
No, there is no need for that.

Reply:
ok, just to clarify here--you're saying it's unnecessary because the ultimate effect of dereferencing the pointer is the same?

Reply:
Something like that...  There is really no benefit of creating a `[1 x i32]` type here.  Using aggregate types allows LLVM IR to express the address computations in a target-independent way (via `getelementptr`), but arrays are always contiguous, so loading the 0th element of `[1 x i32]` is the same as loading from `i32*` that is the address of that element.

Reply:
In fact, in this code
```
define i32 @fred(i32* %a0) {
  %v0 = bitcast i32* %a0 to [1 x i32]*
  %v1 = getelementptr [1 x i32], [1 x i32]* %v0, i32 0, i32 0
  %v2 = load i32, i32* %v1, align 4
  ret i32 %v2
}
```
the use of the array type gets optimized out (in `instcombine`):
```
$ opt -S -instcombine < 1xi32.ll
; ModuleID = '<stdin>'
source_filename = "<stdin>"

define i32 @fred(i32* %a0) {
  %v2 = load i32, i32* %a0, align 4
  ret i32 %v2
}
```

Reply:
thanks for the great explanation @kparzysz-quic !